executor/studentapi.pyx

```diff
@@ -0,0 +1,238 @@
+        
+
+    def is_running(self, action) -> bool:
+        """Returns whether the given function `action` is running in a different thread.
+        
+        Args:
+            action: Python function to check
+        """
+        thread = self.running_actions.get(action.__name__, None)
+        if thread:
+            return thread.is_alive()
+        return False
+
+
+    cpdef get_value(self, str device_id, str param_name):
+        """ 
+        Get a device value. 
+        
+        Args:
+            device_id: string of the format '{device_type}_{device_uid}' where device_type is LowCar device ID and      device_uid is 64-bit UID assigned by LowCar.
+            param_name: Name of param to get. List of possible values are at https://pioneers.berkeley.edu/software/robot_api.html
+        """
+        param_bytes = param_name.encode('utf-8')
+        cdef char* param = param_bytes
+        splits = device_id.split('_')
+        cdef int device_type = int(splits[0])
+        cdef uint64_t device_uid = int(splits[1])
+        cdef param_val_t* param_value = <param_val_t*> PyMem_Malloc(sizeof(param_val_t) * MAX_PARAMS)
+        if not param_value:
+            raise MemoryError("Could not allocate memory to get device value.")
+        cdef param_desc_t* param_desc
+        cdef uint8_t param_idx
+        with nogil:
+            param_desc = get_param_desc(device_type, param)
+            param_idx = get_param_idx(device_type, param)
+            device_read_uid(device_uid, EXECUTOR, DATA, 1 << param_idx, param_value)
+        param_type = param_desc.type.decode('utf-8')
+        if param_type == 'int':
+            ret = param_value[param_idx].p_i
+        elif param_type == 'float':
+            ret = param_value[param_idx].p_f
+        elif param_type == 'bool':
+            ret = bool(param_value[param_idx].p_b)
+        PyMem_Free(param_value)
+        return ret
+
+
+    cpdef void set_value(self, str device_id, str param_name, value):
+        """ 
+        Set a device parameter.
+        
+        Args:
+            device_id: string of the format '{device_type}_{device_uid}' where device_type is LowCar device ID and      device_uid is 64-bit UID assigned by LowCar.
+            param_name: Name of param to get. List of possible values are at https://pioneers.berkeley.edu/software/robot_api.html
+            value: Value to set for the param. The type of the value can be seen at https://pioneers.berkeley.edu/software/robot_api.html
+        """
+        param_bytes = param_name.encode('utf-8')
+        cdef char* param = param_bytes
+        splits = device_id.split('_')
+        cdef int device_type = int(splits[0])
+        cdef uint64_t device_uid = int(splits[1])
+        cdef param_desc_t* param_desc
+        cdef uint8_t param_idx
+        with nogil:
```

Comment:
The purpose of `with nogil` is to allow Python multithreading to use multiple physical cores simultaneously instead of just multiplexing in time. The caveat is that the body must not interact with Python objects.

You'll probably actually incur a penalty for releasing, then reacquiring, the GIL for such short amounts of time very frequently. Unless you're heavily multithreading with CPU-intensive tasks, best to hold onto the GIL.

Reply:
So you think its better if I remove the GIL release? I was originally thinking it could help if the student had multiple actions running at the same time, but I guess its not enough of an increase. A while ago, I tried testing if adding it slowed the code down and if I remember it was slightly slower, maybe like 5-10%.

Reply:
`nogil` is not necessary if actions are implemented with `asyncio` (which we must do anyway if we want backwards-compatbility). `asyncio` is nice because it's single-threaded, meaning GIL contention is not even relevant, since the OS does not handle concurrency. Using `asyncio` for actions, you'll only need two threads: the main thread for `setup`/`main` (synchronous, uses `SIGALRM` to interrupt long-running code), and an auxiliary daemon thread for actions.

Reply:
I ended up removing the GIL release blocks. For now, I don't think I'll be using asyncio but you're right that the time spent to read from shared memory is so small that releasing and reacquiring the lock seems pointless.